kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java

```diff
@@ -66,245 +82,410 @@ String filterDescriptor() {
+                    }
+                });
             }
-            var future = stage.toCompletableFuture();
-            boolean defer = !future.isDone();
-            var maybeDeferred = defer ? handleDeferredStage(ctx, future, decodedFrame) : future;
-            var execute = executeWrite(ctx, decodedFrame, filterContext, maybeDeferred);
-            var maybeDeferredCompleted = defer ? handleDeferredWriteCompletion(ctx, execute) : execute;
-            return maybeDeferredCompleted.thenApply(filterResult -> null);
         }
         else {
-            if (!(msg instanceof OpaqueRequestFrame)
-                    && msg != Unpooled.EMPTY_BUFFER) {
+            if (!(msg instanceof OpaqueRequestFrame) && msg != Unpooled.EMPTY_BUFFER) {
                 // Unpooled.EMPTY_BUFFER is used by KafkaProxyFrontendHandler#closeOnFlush
-                // but otherwise we don't expect any other kind of message
-                LOGGER.warn("Unexpected message writing to upstream: {}", msg, new IllegalStateException());
+                // but, otherwise we don't expect any other kind of message
+                throw new IllegalStateException("Unexpected message writing to upstream: " + msg);
             }
             ctx.write(msg, promise);
+        }
+    }
+
+    private CompletableFuture<Void> readDecodedResponse(DecodedResponseFrame<?> decodedFrame) {
+        var filterContext = new InternalFilterContext(decodedFrame);
+
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug("{}: Dispatching upstream {} response to filter {}: {}",
+                    channelDescriptor(), decodedFrame.apiKey(), filterDescriptor(), decodedFrame);
+        }
+        var stage = invoker.onResponse(decodedFrame.apiKey(), decodedFrame.apiVersion(),
+                decodedFrame.header(), decodedFrame.body(), filterContext);
+        if (stage == null) {
+            if (LOGGER.isWarnEnabled()) {
+                LOGGER.warn("{}: Filter{} for {} response unexpectedly returned null. This is a coding error in the filter. Closing connection.",
+                        channelDescriptor(), filterDescriptor(), decodedFrame.apiKey());
+            }
+            closeConnection();
+            return CompletableFuture.completedFuture(null);
+        }
+        var future = stage.toCompletableFuture();
+        boolean defer = !future.isDone();
+        var maybeDeferred = defer ? handleDeferredStage(decodedFrame, future) : future;
+        var filterHandled = maybeDeferred
+                .thenApply(FilterHandler::validateFilterResultNonNull)
+                .thenApply(fr -> handleResponseFilterResult(decodedFrame, fr))
+                .exceptionally(t -> handleFilteringException(t, decodedFrame));
+        var maybeDeferredCompleted = defer ? handleDeferredReadCompletion(filterHandled) : filterHandled;
+        return maybeDeferredCompleted.thenApply(responseFilterResult -> null);
+    }
```

Comment:
Personally I find the logic easier to follow if I add, `defer` as an arg to `handleDeferredStage` and`handleDeferredReadCompletion`. If its false simply return the passed in future.

Passing `defer` seems like a bit of a smell, but we need to make consistent decisions between `handleDeferredStage` and `handleDeferredReadCompletion`so we need to decide once if we are defering or not.

That means this block ends up looking like 
```suggestion
boolean defer = !future.isDone();
        var filterHandled = handleDeferredStage(decodedFrame, future, defer)
                .thenApply(FilterHandler::validateFilterResultNonNull)
                .thenApply(fr -> handleRequestFilterResult(decodedFrame, promise, fr))
                .exceptionally(t -> handleFilteringException(t, decodedFrame));
        return handleDeferredWriteCompletion(filterHandled, defer)
                .thenApply(filterResult -> null);
```

Reply:
I went a slightly different direction in https://github.com/k-wall/kroxylicious/pull/8

Reply:
I liked the idea, it is now incorporated into the PR.  I kept it a separate commit.